### DISCUSSION TITLE
Guide for typing custom Fields

### DISCUSSION THREAD

**Discussion by martinlehoux** [typeddjango/django-stubs#2170] [Common issues and solutions]
Hello!

I'm struggling to make a pattern work, but I think it's a common pattern in Django. I would like help to implement this correctly, then I can add it as a FAQ item.

My question is : how can I implement the correct typing for a custom model Field in Django? 

I have a simple wrapper around postgres `DateRange` field
```py
class EnhancedDateRangeField(DateRangeField):
    range_type = EnhancedDateRange
    form_field = EnhancedDateRangeFormField

    def from_db_value(self, value: Range[date] | None, expression, connection) -> EnhancedDateRange | None:
        if value is None:
            return None

        return EnhancedDateRange(lower=value.lower, upper=value.upper, bounds=value.bounds)
```
I also have this typing override
```pyi
class EnhancedDateRangeField(DateRangeField):
    _pyi_private_set_type: EnhancedDateRange
    _pyi_private_get_type: EnhancedDateRange
    _pyi_lookup_exact_type: EnhancedDateRange
    range_type: EnhancedDateRange  # type: ignore[assignment]
    def __get__(self, *args, **kwargs) -> EnhancedDateRange: ...  # type: ignore[no-untyped-def, override]
```

1. As you can expect, it works well until I tried to make a `date_range = EnhancedDateRangeField(null=True)`. At this point, I get `EnhancedDateRangeField is nullable but its generic get type parameter is not optional`
2. I tried to replace all `EnhancedDateRange` in my typing override with a typevar `TEnhanced = TypeVar("TEnhancedDateRange", bound=EnhancedDateRange | None)`, but then it requires to add type annotations to all usage of `EnhancedDateRangeField`
3. I also tried to remove the typing overrides, but then the field is `DateRange`

**Comment by Viicos** [comment]
https://github.com/typeddjango/django-stubs/pull/1900 would really help, however to be even more user friendly with subclasses, it will require some work on the mypy side. I'll see if I can make some progress here
